### PR TITLE
Run modules without recursion for better testability.

### PR DESCRIPTION
**Pull request by justinemarie**
I also ran clang-format resulting in some lines-that-look-edited that really weren't.

### CODE REVIEW

File: core/gate.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#ifndef BESS_CORE_GATE_H
```
Comment (barath) [inline core/gate.h]: Add a trailing _ to the header guard.

File: core/task.cc [hunk 2]
```diff
@@ -7,21 +7,43 @@
...
-  return t->m->RunTask(t->arg);
+  struct task_result ret = t->m->RunTask(t->arg);
+
+  // Depth first goes through all pending modules and services
+  while (ctx.gates_pending()) {
+    struct gate_task task = ctx.pop_ogate_and_packets();
+    gate *ogate = task.gate;
+    struct pkt_batch *next_packets = &(task.batch);
+
+#if TRACK_GATES
```
Comment (barath) [inline core/task.cc]: Could you add a comment saying what this does?

File: core/task.cc [hunk 3]
```diff
@@ -7,21 +7,43 @@
...
+
+  return ret;
 }
 
 struct task *task_create(Module *m, void *arg) {
   struct task *t;
 
   t = (struct task *)mem_alloc(sizeof(*t));
-  if (!t) return nullptr;
+  if (!t)
```
Comment (barath) [inline core/task.cc]: Add braces (here and for other ifs)?

File: core/task.cc [hunk 4]
```diff
@@ -112,9 +136,11 @@ void assign_default_tc(int wid, struct task *t) {
 static int get_next_wid(int *wid) {
   static int rr_next = 0;
 
-  if (num_workers == 0) return -1;
+  if (num_workers == 0)
+    return -1;
 
-  while (!is_worker_active(rr_next)) rr_next = (rr_next + 1) % MAX_WORKERS;
+  while (!is_worker_active(rr_next))
```
Comment (barath) [inline core/task.cc]: Add braces?

File: core/worker.h [hunk 5]
```diff
@@ -112,6 +120,28 @@ class Worker {
     return igate_stack_[stack_depth_];
   }
 
+  // Store gate+packets into tasks for worker to service.
```
Comment (barath) [inline core/worker.h]: Maybe mention in the comment "Returns true upon success."

File: core/task.cc (hunk 2 above)
Comment (melvinw) [inline core/task.cc]: No real need. This gets dropped by #128, which adds documentation somewhere else (unless we care about the commit history)

**Comment by justinemarie** [comment]
Which is not really needed? TRACK GATWS?

**Comment by melvinw** [comment]
Nope, TRACK_GATES is fine. I meant the comment Barath was asking for

File: core/gate.h (hunk 1 above)
Comment (sangjinhan) [inline core/gate.h]: Also, all other header files in the core directory does not have "CORE" in the header guard. Not sure whether we should, but let's keep them consistent at least for now.

**Comment by justinemarie** [comment]
Okay, added requested changes.

**Comment by sangjinhan** [comment]
Merged, thanks!

**Pull request merged**